### ISSUE TITLE
BUG: Request logger instances hit maxListeners when highly concurrent.

### ISSUE THREAD

**Issue by didebeach** [FormidableLabs/express-winston-middleware#3]
May I'm wrong, but I would recommend to Review the lines arround 70 in the index.js file. If I'm not wrong, there is a bug that creates always a new loginstance on every request.

My proposal would be something with a singleton pattern. 
Just an idea. inside of the request: function (opts, baseMeta) {

    var staticLog;

    return function (req, res, next) {
      // Create logger and attach to locals.
      if (staticLog == null) {
        staticLog =  new Log(opts, baseMeta);
      }

      var log = res.locals._log = staticLog;
      var _end = res.end;

// Add request

**Comment by ryan-roemer** [comment]
@didebeach -- The request logger in https://github.com/FormidableLabs/express-winston-middleware/blob/master/index.js#L73 is one-off per-request with the intention that a single request owns a single logger and can do things like mutate the metadata / etc. for the logger.

Since we can easily have concurrent requests, having a static logger here would have  different requests stomp all over each other.

A singleton logger can be manually created a la https://github.com/FormidableLabs/express-winston-middleware#logger

**Comment by ryan-roemer** [comment]
And notice here -- https://github.com/FormidableLabs/express-winston-middleware/blob/master/index.js#L76-L98 -- we're actively _mutating_ the logger instance with metadata for this specific request...

**Comment by ryan-roemer** [comment]
And I guess rounding out how I use the logger middleware --

1. I create a single logger for the entire app, usually attaching to `app._log` or something for global state / logs.
2. For each request, in various controllers, etc., I (a) add specific metadata for eventual logging like `res.locals._log.addMeta({ dbRequestTime: numMs });` and (b) sometimes do extra logging calls like `res.locals._log.info("HI")`

So, that's how I usually balance a single global logger vs. per-request individual logger instances...

**Comment by didebeach** [comment]
Hi ryan
thank you for your feedback. My idea of the singleton was coming out after searching for my current problem related to "NodeJS : warning: possible EventEmitter memory leak detected. 11 listeners added. Use emitter.setMaxListeners() to increase Limit"  (http://stackoverflow.com/questions/13983204/nodejs-warning-possible-eventemitter-memory-leak-detected-11-listeners-added)
I had constant a message that I'm running in a possible Memory leak with the solution, and even if I put my maxListener to 1000 I'm still running into the Problem.

May I did not correct use the Middleware and you can teach me how to use it correct. I understand that you like to have an instance for every request, but my question would be: When you will Close the instance?
It Looks for me that we just adding the number of instances on every request, but we never decrease the number by deleting instances.

**Comment by ryan-roemer** [comment]
`res.locals` are scoped to the request and should be cleaned up along with everything else in the responses. So `res.locals` goes out of scope at the end of the response and `res.locals._log` should be eventually released there. See, e.g., http://expressjs.com/api.html#res.locals

> An object that contains response local variables scoped to the request, and therefore available only to the view(s) rendered during that request / response cycle (if any).

Out of curiosity, does your app _not_ run into the `maxListener` error when using the singleton approach? If so, maybe there is some hanging reference somewhere in the middleware, and we can potentially hunt that down...

**Comment by didebeach** [comment]
Yes, I understand your comment with automatic unloading and I do not hit the maxListener (event with default settings to 10) with the singleton approach. So may your point to having hanging instances may be correct and I will have to double check that tomorrow again if I'm back in the office.

Indepencent, may once approach to bring both worlds together can be solved like this?
var log = res.locals._log = (app._log ||  new Log(opts, baseMeta) );
In this case it would make sense that I can have a singleton (by using the app._log instance) or without having this instance we creating everytime a new one. I would prefer to have may an parameter that allows me to not using the app._log approach, instead I would just set the ooptions for the logger to (singleton: true -->false by default) and this would not change your style of programming but would be helpfull for me if I like to have a singleton.

But may I first have to understand the benefit of having for every reguest a single logger instance. 
As fare as I understand we talking about a middleware that always will be called automatically by express on every request. My understanding would be that in this case we always using the same initializer and there is no benefit of having for every request a fresh instance.

Sorry, I'm not having that much deep experience with Javascript and like to learn from others their approach of solving problems.

**Comment by ryan-roemer** [comment]
@didebeach -- Let me try a branch that will explicitly delete the logger.

The singleton approach is virtually guaranteed to be wrong because metadata is attached upon request _start_ and then later on response _finish_.  Assuming you have any other requests / responses do anything in the meantime (while waiting on IO), the metadata is going to be an unpredictable mismash of several requests....

**Comment by ryan-roemer** [comment]
@didebeach -- Can you create a reproduction of your error / point me to a chunk of code to run and see this error in action? It's taking too long for me to try and solve it without seeing it myself.

Also, for the reproduction code sample, can you also let me know the following information about your environment:

* Node version
* OS

Thanks!

**Comment by didebeach** [comment]
Using the followin Development environment
Windows Server 2012 R2 standard,
node v0.10.36

and a small sample app:
/*
* used the following dependencies
*
 "debug": "2.1.3",
 "express": "4.12.3",
 "express-winston-middleware": "0.0.4",
 "winston": "0.9.0"

 after pointing to the localhost:3000 and pressing 11times refresh, you can find the warning....

(node) warning: possible EventEmitter memory leak detected. 11 listeners added. Use emitter.setMaxListeners() to increase limit.
Trace
    at addListener (events.js:160:15)
    at Logger.add (c:\Users\ddeg\WebstormProjects\winstonTest\node_modules\winston\lib\winston\logger.js:456:12)
    at exports.Logger.handleExceptions (c:\Users\ddeg\WebstormProjects\winstonTest\node_modules\winston\lib\winston\logger.js:62:12)
    at Array.forEach (native)
    at new exports.Logger (c:\Users\ddeg\WebstormProjects\winstonTest\node_modules\winston\lib\winston\logger.js:61:24)
    at new Log (c:\Users\ddeg\WebstormProjects\winstonTest\node_modules\express-winston-middleware\index.js:213:15)
    at middleware.request (c:\Users\ddeg\WebstormProjects\winstonTest\node_modules\express-winston-middleware\index.js:73:35)
    at Layer.handle [as handle_request] (c:\Users\ddeg\WebstormProjects\winstonTest\node_modules\express\lib\router\layer.js:82:5)
    at trim_prefix (c:\Users\ddeg\WebstormProjects\winstonTest\node_modules\express\lib\router\index.js:302:13)
    at proto.handle.c (c:\Users\ddeg\WebstormProjects\winstonTest\node_modules\express\lib\router\index.js:270:7
 *
* */

var express = require('express');

var app = express();
var server = require('http').createServer(app);

var winston = require('winston');
var winstonMiddleware = require("express-winston-middleware");
var logOptions =
{
  transports: [
    new (winston.transports.Console)(
      {
        json: false,
        level: 'silly',
        colorize: true
      }
    )
  ]
};


app.use(winstonMiddleware.request(logOptions, {type: "per-request-log"}));

app.use('/', function(req,res){
  res.end("Hello world");
});

server.listen(
  3000,
  '0.0.0.0',
  function () {
    console.log("Express server listening on port " + this.address().port);
  }
);

**Comment by ryan-roemer** [comment]
@didebeach -- Thanks! I'll try and get this up and running later today...

**Comment by ryan-roemer** [comment]
Confirmed. Easier reproduction:

* `node examples/server.js`
* Keep refreshing http://127.0.0.1:9876/custom-logging

... and will eventually see the WARN message.

Preliminary research indicates an upstream Winston listener bug [#218](https://github.com/winstonjs/winston/issues/218) "Leaking listeners when 'handleExceptions: true' and 'exitOnError: false'"

**Comment by didebeach** [comment]
The exitOnError point me out, that may my initial problem is homemade. May there is a problem with my definition of the logger options.

Neverless I was  trying to check if my optimized version would also work and....

var express = require('express');

var app = express();
var server = require('http').createServer(app);

var winston = require('winston');
var winstonMiddleware = require("express-winston-middleware");
var logOptions =
{
  transports: [
    new (winston.transports.Console)(   )
  ]
};

app.use(winstonMiddleware.request(logOptions, {type: "per-request-log"}));

app.use(function(req, res, next) {
  res.locals._log.addMeta({ middleware1: 1 });
  res.locals._log.info("HI");

  next();
});

app.use(function(req, res, next) {
  res.locals._log.addMeta({ middleware2: 2 });
  res.locals._log.info("HI");

  next();
});


app.use('/', function(req,res){
  res.end("Hello world");
});

server.listen(
  3000,
  '0.0.0.0',
  function () {
    console.log("Express server listening on port " + this.address().port);
  }
);

If I'm looking now for the output with my solution or your solution: looks same...
my version with the hash and only one logger instance per options:
info: HI date=2015-03-31T17:39:56.153Z, env=development, id=m, pid=8740, hostName=Yoga2Pro, type=per-request-log, method=GET, host=localhost:3000, path=/, query=, ip=127.0.0.1, middleware1=1
info: HI date=2015-03-31T17:39:56.155Z, env=development, id=m, pid=8740, hostName=Yoga2Pro, type=per-request-log, method=GET, host=localhost:3000, path=/, query=, ip=127.0.0.1, middleware1=1, middleware2=2
info: request date=2015-03-31T17:39:56.161Z, env=development, id=m, pid=8740, hostName=Yoga2Pro, type=per-request-log, method=GET, host=localhost:3000, path=/, query=, ip=127.0.0.1, middleware1=1, middleware2=2, statusCode=200

your version a logger instance per request:
info: HI date=2015-03-31T17:44:27.211Z, env=development, id=m, pid=632, hostName=Yoga2Pro, type=per-request-log, method=GET, host=localhost:3000, path=/, query=, ip=127.0.0.1, middleware1=1
info: HI date=2015-03-31T17:44:27.215Z, env=development, id=m, pid=632, hostName=Yoga2Pro, type=per-request-log, method=GET, host=localhost:3000, path=/, query=, ip=127.0.0.1, middleware1=1, middleware2=2
info: request date=2015-03-31T17:44:27.225Z, env=development, id=m, pid=632, hostName=Yoga2Pro, type=per-request-log, method=GET, host=localhost:3000, path=/, query=, ip=127.0.0.1, middleware1=1, middleware2=2, statusCode=200

looks very similar to me.......

I will have to fix my declaration but I would still recommend to recycle the winston logger instance and not creating always a new one. Can imagine that this it will be more "expensive" to create a new instance as to reuse it.

Thank you anyway that you point me to the right direction
